src/test/java/BaseTest.java

```diff
@@ -1,11 +1,104 @@
 import io.github.bonigarcia.wdm.WebDriverManager;
+import org.openqa.selenium.By;
+import org.openqa.selenium.WebDriver;
+import org.openqa.selenium.WebElement;
+import org.openqa.selenium.chrome.ChromeDriver;
 import org.openqa.selenium.chrome.ChromeOptions;
-import org.testng.annotations.BeforeSuite;
+import org.openqa.selenium.firefox.FirefoxDriver;
+import org.openqa.selenium.firefox.FirefoxOptions;
+import org.openqa.selenium.interactions.Actions;
+import org.openqa.selenium.remote.DesiredCapabilities;
+import org.openqa.selenium.remote.RemoteWebDriver;
+import org.openqa.selenium.safari.SafariDriver;
+import org.openqa.selenium.safari.SafariOptions;
+import org.openqa.selenium.support.ui.ExpectedConditions;
+import org.openqa.selenium.support.ui.WebDriverWait;
+import org.testng.annotations.*;
+import pages.*;
+
+import java.net.MalformedURLException;
+import java.net.URI;
+import java.net.URL;
+import java.time.Duration;
+import java.time.Instant;
 
 public class BaseTest {
+    private static final ThreadLocal<WebDriver> threadDriver = new ThreadLocal<>();
+    public static WebDriver getThreadDriver(){
+        return threadDriver.get();
+    }
+    public static WebDriver driver = null;
+    public String url = "https://qa.koel.app/";
+    WebDriverWait wait = null;
+    Actions actions = null;
+    BasePage basePage;
+    HomePage homePage;
+    LoginPage loginPage;
+    RegistrationPage registrationPage;
+
+    @BeforeMethod
+    public void setupTest() throws MalformedURLException {
+        threadDriver.set(setupBrowser(System.getProperty("browser")));
```

Comment:
System Property "browser" Not Set: The System.getProperty("browser") call in the setupTest method retrieves the value of the "browser" system property. If this system property is not set, it will return null. Make sure that when you run your tests, you are setting the "browser" property.

`threadDriver.set(setupBrowser(System.getProperty("browser", "chrome")));`